src/Features/CSharp/Portable/CodeMapper/CSharpMapCodeService.NodeToMap.cs

```diff
@@ -0,0 +1,423 @@
+
+        /// <summary>
+        /// The name of the node's identifier.
+        /// </summary>
+        public string IdentifierName => _identifierName.Value;
+
+        public string ToFullString() => Node.ToFullString();
+
+        /// <summary>
+        /// The syntax node wrapped by this class.
+        /// </summary>
+        public readonly SyntaxNode Node;
+
+        /// <summary>
+        /// Gets the scope of the node.
+        /// </summary>
+        public Scope Scope { get; private set; }
+
+        /// <summary>
+        /// Returns a value indicating whether this node is valid replace type of node.
+        /// Replace nodes currently only support Class and Method.
+        /// </summary>
+        public bool IsReplaceScope => Scope == Scope.Method || Scope == Scope.Class;
+
+        /// <summary>
+        /// Creates a new instance of the SourceNode class.
+        /// </summary>
+        /// <param name="node">The syntax node to represent.</param>
+        private NodeToMap(SyntaxNode node, Scope scope)
+        {
+            Node = node;
+            Scope = scope;
+
+            _identifierName = new Lazy<string>(() => GetIdentifierName(Node));
+        }
+
+        /// <summary>
+        /// Checks whether this node exists on a given target node.
+        /// </summary>
+        /// <param name="node">The target node to check.</param>
+        /// <param name="matchingNode">If this node exists on the target node, this is the matching node.</param>
+        /// <returns>True if this node exists on the target node, false otherwise.</returns>
+        public bool ExistsOnTarget(SyntaxNode node, out SyntaxNode? matchingNode)
+        {
+            matchingNode = null;
+
+            if (Scope is Scope.None)
+            {
+                matchingNode = node.DescendantNodesAndSelf()
+                    .Where(n => IsSimpleNode(n) && GetIdentifierName(n) == IdentifierName)
+                    .FirstOrDefault();
+            }
+            else if (IsReplaceScope)
+            {
+                matchingNode = node.DescendantNodesAndSelf()
+                                    .Where(n => IsScopedNode(n, out var scope) && scope == Scope && GetIdentifierName(n) == IdentifierName)
+                                    .FirstOrDefault();
+            }
+
+            return matchingNode is not null;
+        }
+
+        /// <summary>
+        /// Returns a string representation of the node.
+        /// </summary>
+        /// <returns>A string representation of the node.</returns>
+        public override string ToString()
+        {
+            return Node.ToFullString();
+        }
+
+        /// <summary>
+        /// Gets the identifier name of a syntax node.
+        /// </summary>
+        /// <param name="node">The syntax node to get the identifier name for.</param>
+        /// <returns>The identifier name of the syntax node.</returns>
+        private static string GetIdentifierName(SyntaxNode node)
```

Comment:
this is *Extremely* worrying.  It feels like trying to just guess that nodes are the same because of a matchign identifier.  this needs to eb fleshed out as to why this is correct/desriable.